jlm/llvm/opt/alias-analyses/RegionAwareModRefSummarizer.cpp

```diff
@@ -712,299 +646,619 @@ RegionAwareModRefSummarizer::CreateCallGraph(const rvsdg::RvsdgModule & rvsdgMod
+{
+  const auto & region = *lambda.subregion();
+  const auto regionModRefSet = AnnotateRegion(region, lambda);
+  const auto lambdaModRefSet = ModRefSummary_->GetOrCreateSetForNode(lambda);
+  AddModRefSimpleConstraint(regionModRefSet, lambdaModRefSet);
+}
+
+ModRefSetIndex
+RegionAwareModRefSummarizer::AnnotateRegion(
+    const rvsdg::Region & region,
+    const rvsdg::LambdaNode & lambda)
+{
+  const auto regionModRefSet = ModRefSummary_->CreateModRefSet();
+
+  for (auto & node : region.Nodes())
   {
-    AnnotateCall(simpleNode, regionSummary);
+    if (auto structuralNode = dynamic_cast<const rvsdg::StructuralNode *>(&node))
+    {
+      const auto nodeModRefSet = AnnotateStructuralNode(*structuralNode, lambda);
+      AddModRefSimpleConstraint(nodeModRefSet, regionModRefSet);
+    }
+    else if (auto simpleNode = dynamic_cast<const rvsdg::SimpleNode *>(&node))
+    {
+      const auto nodeModRefSet = AnnotateSimpleNode(*simpleNode, lambda);
+      if (nodeModRefSet)
+        AddModRefSimpleConstraint(*nodeModRefSet, regionModRefSet);
+    }
+    else
+    {
+      JLM_UNREACHABLE("Unhandled node type!");
+    }
   }
-  else if (is<MemCpyOperation>(&simpleNode))
+
+  // Check if this region has any Non-Reentrant Allocas. If so, block them
+  if (const auto it = Context_->NonReentrantAllocas.find(&region);
+      it != Context_->NonReentrantAllocas.end() && ENABLE_NON_REENTRANT_ALLOCA_BLOCKLIST)
   {
-    AnnotateMemcpy(simpleNode, regionSummary);
+    JLM_ASSERT(ModRefSummary_->GetModRefSet(regionModRefSet).IsEmpty());
+    AddModRefSetBlocklist(regionModRefSet, it->second);
   }
+
+  return regionModRefSet;
+}
+
+ModRefSetIndex
+RegionAwareModRefSummarizer::AnnotateStructuralNode(
+    const rvsdg::StructuralNode & structuralNode,
+    const rvsdg::LambdaNode & lambda)
+{
+  const auto nodeModRefSet = ModRefSummary_->GetOrCreateSetForNode(structuralNode);
+
+  for (auto & subregion : structuralNode.Subregions())
+  {
+    const auto subregionModRefSef = AnnotateRegion(subregion, lambda);
+    AddModRefSimpleConstraint(subregionModRefSef, nodeModRefSet);
+  }
+
+  return nodeModRefSet;
+}
+
+std::optional<ModRefSetIndex>
+RegionAwareModRefSummarizer::AnnotateSimpleNode(
+    const rvsdg::SimpleNode & simpleNode,
+    const rvsdg::LambdaNode & lambda)
+{
+  if (is<LoadOperation>(&simpleNode))
+    return AnnotateLoad(simpleNode, lambda);
+
+  if (is<StoreOperation>(&simpleNode))
+    return AnnotateStore(simpleNode, lambda);
+
+  if (is<AllocaOperation>(&simpleNode))
+    return AnnotateAlloca(simpleNode);
+
+  if (is<MallocOperation>(&simpleNode))
+    return AnnotateMalloc(simpleNode);
+
+  if (is<FreeOperation>(&simpleNode))
+    return AnnotateFree(simpleNode, lambda);
+
+  if (is<MemCpyOperation>(&simpleNode))
+    return AnnotateMemcpy(simpleNode, lambda);
+
+  if (is<CallOperation>(&simpleNode))
+    return AnnotateCall(simpleNode, lambda);
+
+  return std::nullopt;
 }
```

Comment:
Can we have a fail-safe again here that checks for memory inputs/outputs of the node?